File: AtomykPlay/app/src/main/java/com/atomykcoder/atomykplay/function/FetchLyrics.java
package com.atomykcoder.atomykplay.function;

import android.os.Bundle;
import android.view.View;
import android.widget.ProgressBar;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.io.IOException;
import java.util.ArrayList;

public class FetchLyrics {

    /**
     * Toggles Progressbar to View.VISIBLE
     * @param progressBar
     */
    public void onPreExecute(ProgressBar progressBar) {
        // pre-execute code goes here
        progressBar.setVisibility(View.VISIBLE);
    }

    /**
     * Fetches List of All Songs that match the given query
     * @param query query that needed to be searched (a song name + artist name)
     * @return returns a bundle with song titles, samplelyrics and urls
     */
    public Bundle fetchList (String query) {
        Bundle bundle = new Bundle();
        Element titleLink;
        Element lyricsLink;
        String lyrics;
         ArrayList<String> titles = new ArrayList<>();
         ArrayList<String> sampleLyrics = new ArrayList<>();
         ArrayList<String> urls = new ArrayList<>();
        try {
            final Document document = Jsoup.connect("https://www.megalobiz.com/search/all?qry=" + query).get();
            Elements titleElements = document.select("div.pro_part.mid");
            Elements lyricsElements = document.select("div.details.mid");

            // Retrieve 10 Items from the list
            for ( int i = 0; i < 20; i++ ) {
                //get Title and urls
                if(titleElements != null && !titleElements.isEmpty()) {
                    titleLink = titleElements.get(i).select("a").first();
                    if (titleLink != null) {
                            titles.add(titleLink.text());
                        urls.add(titleLink.attr("href"));
                    }
                }
                //get sample Lyrics
                if(lyricsElements != null && !lyricsElements.isEmpty()) {
                    lyricsLink = lyricsElements.get(i).select("div").get(2);
                    lyricsLink = lyricsLink.select("span").first();
                    if (lyricsLink != null) {
                        lyrics = MusicHelper.splitLyricsByNewLine(lyricsLink.text());
                        sampleLyrics.add(lyrics);
                    }
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        // put all the data in a bundle nicely wrapped
        bundle.putStringArrayList("titles", titles);
        bundle.putStringArrayList("sampleLyrics", sampleLyrics);
        bundle.putStringArrayList("urls", urls);

        //and return the bundle
        return bundle;
    }

    /**
     * Fetches the song associated with given weblink
     * @param href weblink associated with a song
     * @return returns the Time Stamps of the song
     */
    public String fetchTimeStamps (String href) {
        Element lyrics = null;
        try {
            final Document lyricsDocument = Jsoup.connect("https://www.megalobiz.com" + href).get();
            for (Element div : lyricsDocument.select("div.lyrics_details.entity_more_info")) {
                lyrics = div.select("span").first();
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        if (lyrics != null)
            return lyrics.text();
        else
            return "";
    }

    /**
     * Toggles Progressbar to View.GONE
     * @param progressBar
     */
    public void onPostExecute(ProgressBar progressBar) {
        progressBar.setVisibility(View.GONE);
    }

    //region get duration method maybe for use later...
//    private String getLength(String songName) {
//        String o = "[00:00.00]";
//        Log.i("match", songName);
//        Pattern _pattern = Pattern.compile("\\[\\d\\d:\\d\\d.\\d\\d\\]");
//        Matcher _matcher = _pattern.matcher(songName);
//        if(_matcher.find()) {
//            o = songName.substring(_matcher.start(), _matcher.end());
//        }
//        Log.i("match", "Match : " +  o);
//        return o;
//    }
    //endregion.......
}

File: AtomykPlay/app/src/main/java/com/atomykcoder/atomykplay/helperFunctions/FetchLyrics.kt
package com.atomykcoder.atomykplay.helperFunctions

import android.os.Bundle
import android.view.View
import android.widget.ProgressBar
import com.atomykcoder.atomykplay.interfaces.ApiService
import okhttp3.OkHttpClient
import okhttp3.ResponseBody
import org.jsoup.Jsoup
import org.jsoup.nodes.Element
import retrofit2.Retrofit
import retrofit2.converter.scalars.ScalarsConverterFactory
import java.io.IOException
import java.util.concurrent.TimeUnit

class FetchLyrics {

    private val retrofit: Retrofit by lazy {
        Retrofit.Builder()
            .baseUrl("https://www.megalobiz.com/")
            .client(createOkHttpClient())
            .addConverterFactory(ScalarsConverterFactory.create())
            .build()
    }

    private val apiService: ApiService by lazy {
        retrofit.create(ApiService::class.java)
    }

    /**
     * Create and configure OkHttpClient with logging interceptor
     */
    private fun createOkHttpClient(): OkHttpClient {
        return OkHttpClient.Builder()
            .readTimeout(45, TimeUnit.SECONDS)
            .callTimeout(45, TimeUnit.SECONDS)
            .build()
    }

    /**
     * Toggles Progressbar to View.VISIBLE
     */
    fun onPreExecute(progressBar: ProgressBar) {
        progressBar.visibility = View.VISIBLE
    }

    /**
     * Fetches List of All Songs that match the given query
     *
     * @param name query that needed to be searched (song name)
     * @param artistName query that needed to be searched (artist name)
     * @return returns a bundle with song titles, sample lyrics, and URLs
     */
    suspend fun fetchList(name: String?, artistName: String?): Bundle? {
        val bundle = Bundle()
        val titles = ArrayList<String>()
        val sampleLyrics = ArrayList<String>()
        val urls = ArrayList<String>()
        try {
            val response = apiService.search("$name $artistName")

            val document = Jsoup.parse(response.string())

            val titleElements = document.select("div.pro_part.mid")
            val lyricsElements = document.select("div.details.mid")

            if (titleElements != null) {
                // Retrieve items from the list
                for (i in 0..9) {
                    // Get Title and URLs
                    if (!titleElements.isEmpty()) {
                        val titleLink = titleElements[i].select("a").first()
                        if (titleLink != null) {
                            titles.add(titleLink.text())
                            urls.add(titleLink.attr("href"))
                        }
                    }
                    // Get sample Lyrics
                    if (lyricsElements != null && !lyricsElements.isEmpty()) {
                        var lyricsLink = lyricsElements[i].select("div")[2]
                        lyricsLink = lyricsLink.select("span").first()
                        if (lyricsLink != null) {
                            val lyrics = MusicHelper.splitLyricsByNewLine(lyricsLink.text())
                            sampleLyrics.add(lyrics)
                        }
                    }
                }
            } else {
                return null
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }

        // Put all the data in a bundle nicely wrapped
        bundle.putStringArrayList("titles", titles)
        bundle.putStringArrayList("sampleLyrics", sampleLyrics)
        bundle.putStringArrayList("urls", urls)

        // And return the bundle
        return bundle
    }

    /**
     * Fetches the song associated with the given weblink
     *
     * @param href weblink associated with a song
     * @return returns the Time Stamps of the song
     */
    suspend fun fetchTimeStamps(href: String): String {
        var lyrics: Element? = null
        try {
            val response: ResponseBody = apiService.getSong(href)
            val lyricsDocument = Jsoup.parse(response.string())
            for (div in lyricsDocument.select("div.lyrics_details.entity_more_info")) {
                lyrics = div.select("span").first()
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return lyrics?.text() ?: ""
    }

    /**
     * Toggles Progressbar to View.GONE
     */
    fun onPostExecute(progressBar: ProgressBar) {
        progressBar.visibility = View.GONE
    }
}
